Add custom data type support into Modbus climate

vzahradnik opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Climate configuration was moved from the platform directly into the `Modbus` integration. This change was required to meet the 
latest [architecture requirements](https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md) for Home Assistant.

### Before
```yaml
climate:
  - platform: modbus
    name: Watlow F4T
    hub: hub1
    slave: 1
    data_type: uint
    data_count: 1
    scale: 0.1
    offset: 0
    precision: 1
    max_temp: 30
    min_temp: 15
    temp_step: 1
    target_temp_register: 2782
    current_temp_register: 27586
```

### After
```yaml
# Example configuration.yaml
modbus:
  - name: hub1
    type: tcp
    host: 127.0.0.1
    port: 5020

    climates:
      - name: Watlow F4T
        slave: 1
        data_type: uint
        data_count: 1
        scale: 0.1
        offset: 0
        precision: 1
        max_temp: 30
        min_temp: 15
        temp_step: 1
        target_temp_register: 2782
        current_temp_register: 27586
```

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This commit adds support for custom data type into Modbus climate,
and improves error handling.

Currently if a user sets `data_count` into a value other than 1, 2 or 4 bytes,
code throws an exception because it didn't support the parsing.

New behavior is to notify user that parsing was not successful,
and to inform him to specify a custom data structure to use.

This behavior will be the same as used in Modbus sensor.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
modbus:
  - name: hub1
    type: tcp
    host: 127.0.0.1
    port: 5020

    climates:
      name: Watlow F4T
      slave: 1
      data_type: custom
      structure: ">f"
      data_count: 1
      target_temp_register: 2782
      current_temp_register: 27586
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: home-assistant/home-assistant.io#12250

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

springstan commented on homeassistant/components/modbus/climate.py:
```diff
@@ -66,48 +68,69 @@
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
     """Set up the Modbus Thermostat Platform."""
-    name = config[CONF_NAME]
```
Please do not remove the declaration of these local variables. They make the code easier to read.

springstan commented on homeassistant/components/modbus/climate.py:
```diff
@@ -66,48 +68,69 @@
     hub_name = config[CONF_HUB]
     hub = hass.data[MODBUS_DOMAIN][hub_name]
 
+    data_types = {
```
Please create a constant for this dictionary and move it to the other constants at the top.

springstan commented on homeassistant/components/modbus/climate.py:
```diff
@@ -66,6 +74,7 @@
         vol.Optional(CONF_MAX_TEMP, default=35): cv.positive_int,
         vol.Optional(CONF_MIN_TEMP, default=5): cv.positive_int,
         vol.Optional(CONF_STEP, default=0.5): vol.Coerce(float),
+        vol.Optional(CONF_STRUCTURE, default=">f"): cv.string,
```
Please create a constant for this default value e.g. `DEFAULT_STRUCTURE` or `DEFAULT_STRUCTURE_PREFIX`

springstan commented on homeassistant/components/modbus/climate.py:
```diff
@@ -89,6 +98,28 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+
+    if data_type != DATA_TYPE_CUSTOM:
+        try:
+            structure = f">{DEFAULT_STRUCT_FORMAT[data_type][count]}"
```
Use the constant here as well.

vzahradnik replied:
What part of the expression should I put into constant? Right now, a struct format is selected from an array based on data type and count values.

janiversen replied:
Are we really asking the user to add something like “>f” (default for structure), would be nice if it was something simpler.

vzahradnik replied:
@janiversen, `CONF_STRUCTURE` is an optional parameter, which is not used under normal circumstances. In most cases, user gives `data_count` and `data_type` parameters, and based on them we select a proper structure matching data type and number of bytes.

We list only few common types there, and sometimes data aren't parsed correctly. Therefore the `struct` option. Users who use this option, should be well aware of what they're doing, and they're given full control. From my point of view, this is acceptable.

janiversen replied:
I suspected that, I will take a closer look at the documentation.

springstan commented on homeassistant/components/modbus/climate.py:
```diff
@@ -89,6 +98,28 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+            structure = f">{DEFAULT_STRUCT_FORMAT[data_type][count]}"
+        except KeyError:
+            _LOGGER.error(
+                "Unable to detect data type for %s sensor, try a custom type", name,
```
Not very familar with this integration but shouldn't we name that it could not be found because of `count`?
Because invalid `data_type`s are filter via the config validation beforehand.

vzahradnik replied:
Yes, we can change that error message. I'll do that.

vzahradnik commented:
@springstan thanks for the comments. I will postpone my work until PR #32557 gets merged. It refactored most of the Modbus integration, and my code will need to be adapted for that.

janiversen commented:
I think that is a good idea. Be aware I am working on changing the tests, so that there a common part (basically run_test) and each entity only contains the actual test

vzahradnik commented:
@janiversen yeah, I saw your comments in that other PR. Thanks for the heads up.

janiversen reviewed (changes requested):
This PR needs to be updated to our new standard. Once updated I will make a more in depth review, but in general it looks good.

janiversen commented:
The reworked test harness is now merged, so no excuses for not writing test cases :-)

vzahradnik commented:
@janiversen, I rebased the code against latest `dev`. Also, I moved the `DEFAULT_STRUCT_FORMAT` into the `const.py` file, so that it can be shared between `climate` and `sensor`. Could you take a look? Thanks!

janiversen reviewed (commented):
I had hoped you would have added just a couple of tests. We need that to get a higher (than no-score) on our integration.

My question about >f is just a question, not a request that you do it differently.

janiversen commented:
Regarding the pytest error, pull newest dev and rebase then it is gone.

vzahradnik commented:
@janiversen just today I checked the integration quality scale requirements. If they're still up to date (and they should be), we meet almost all requirements to get the *silver* rating. Unit tests are not a requirement there.

What we don't meet is following:
- Handles internet unavailable. Log a warning **once** when unavailable, log once when reconnected.
- Handles device/service unavailable. Log a warning **once** when unavailable, log once when reconnected.

If we manage to fix the logging, this integration could be certified. At the moment, `pymodbus` warning is logged on every failed attempt, e.g., a connection is lost.

And regarding the tests, I didn't have time this week. I'm glad that I found some time to go through all my pull requests.

janiversen reviewed (approved):
Reviewed this version of the code, and it is ready (from my side) to be merged.

vzahradnik commented:
Still working on it.

vzahradnik commented:
@janiversen please check this PR. Rebased against latest dev and re-tested.

vzahradnik commented:
Still working on it.

vzahradnik commented:
This PR is still pending review from maintainers. I can, however, rebase the change to find out it still works with latest Home Assistant code.

janiversen commented:
??? as one of the code owners and maintainer I approved the code 13 of April. You have a number of good Prs which should have been merged for a long time, I hope we can soon get it done.

Let me know if I can do anything?

MartinHjelmare commented on homeassistant/components/modbus/climate.py:
```diff
@@ -67,7 +72,8 @@
         vol.Optional(CONF_MIN_TEMP, default=5): cv.positive_int,
         vol.Optional(CONF_STEP, default=0.5): vol.Coerce(float),
-        vol.Optional(CONF_UNIT, default="C"): cv.string,
+        vol.Optional(CONF_STRUCTURE, default=DEFAULT_STRUCTURE_PREFIX): cv.string,
```
I think we've already talked about this for another modbus platform:
We no longer allow new integrations to use a platform config section or old integrations to do changes to the platform config section. See https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md.

Please refactor this integration to have the config schema in the base integration module and the config yaml section under the top level integration domain key.

vzahradnik replied:
@MartinHjelmare Sure, I can do that. But this is not a new integration. If I change that, we'll break backwards compatibility. Please let me know if I should proceed.

vzahradnik replied:
OK, I read the document. Basically, if I want to add a new config entry, the whole config section needs to be reworked.

@MartinHjelmare in PR #33642 you suggested to adapt the config entry in a way to not break compatibility. However, in this case, the compatibility will be broken. Perhaps it is time to go through the whole Modbus integration and move the configuration into a single place to meet your new coding guidelines once for all.

This change would involve changes inside main Modbus section (especially moving hub configuration into a separate subentry).

As an alternative, we can be more cautious, and change only what's needed for Climate and Cover integration. @janiversen, @MartinHjelmare what's your suggestion on this?

For now, I will prepare changes so that they'll break only Climate config.

MartinHjelmare replied:
We should follow the cover platform example.

janiversen replied:
Thjs is existing, so we should be very careful not to break any existing configurations.

I would really prefer to not mame breaking changes in climate, but wait a bit (as discussed earlier) and then change the whole of modbus.

vzahradnik replied:
@janiversen the Home Assistant guidelines affect also existing integrations. According to the document, I can't add a new config entry under platform, which is deprecated now.

This change affects only the Climate entity. Perhaps I could check if the code is used elsewhere first, and if not, I think we should get this merged.

MartinHjelmare replied:
Until the config structure is changed, no config option changes are allowed in this platform.

vzahradnik replied:
OK, so do we keep this PR open until we do a bigger refactor of the whole Modbus?

janiversen replied:
that is not fair against you, let us get it merged and deal with any side effects later

vzahradnik replied:
Actually, I'm OK with that. The other changes are more important for me. This one is different - I ported what was already implemented in Modbus sensor into the Climate. It is handy feature to have, but it can wait.

However, if we decide to keep this open, somebody should ideally put a flag here, so that this PR won't be closed for being stale.

Let me know what you think it's best, and we'll go from there...

MartinHjelmare replied:
If we're not planning to merge this soon, we should close the PR. It can be reopened later when we want to merge it.

vzahradnik replied:
@janiversen what do you suggest? I am OK with both scenarios - postponing the change, or making necessary changes in the config to meet the requirements.

If we postpone, at least we should talk about the timeframe when we're going to introduce the necessary breaking changes to support new way of configuration.

vzahradnik replied:
@MartinHjelmare I reworked the `Climate` configuration to be under the `Modbus` component (the same way as used for the `Cover`, and currently `Light` and `Fan`).

Please add the `breaking-change` label. Documentation was already updated.

Thanks!

vzahradnik commented:
@MartinHjelmare sorry, I missed that comment. Yes, I also want my PRs to get merged. Just I don't have a time to work on them at this moment. I plan to review all of them when I finally get some free time.

Thanks, I will let you know if there's anything.

janiversen commented:
Any idea when you will continue working on this PR. We have a issue open on climate, but I do not want to make your work more difficult by adding a new PR for the same code.

vzahradnik commented:
> Any idea when you will continue working on this PR. We have a issue open on climate, but I do not want to make your work more difficult by adding a new PR for the same code.

@janiversen it looks like I'll need to rework the way how the Climate component is configured. If you want to work on the issue, go ahead. I will deal with your changes, and adapt my code accordingly.

janiversen commented:
OK, I will continue first by adding a couple of simple tests (I am in the process of adding tests for all devices), and then solve the bug, which I think is easy.

I will comment on the other PR regarding the config changes.

vzahradnik commented:
@janiversen OK, thanks! Today I also reworked how config for the `cover` is handled. If @MartinHjelmare approves the way how it's done, I will implement the same way also here to meet the HA requirements:

https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md

janiversen commented on homeassistant/components/modbus/climate.py:
```diff
@@ -90,6 +96,30 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+        _LOGGER.error(
+            "Structure size (%d bytes) mismatch registers count (%d words)", size, count
+        )
+        return
```
We have test cases for sensor, could you please add a couple of cases testing the structure. Remark the reason for the test cases is not to test your code, but to secure a future change does not break it.

Thanks in advance, I think the new much simpler way of adding test cases should make it easy to do, but just call if you need help.

vzahradnik replied:
I will try to add some tests in the upcoming days. And sure, I will contact you if I need some help.

My experience with async tests in Python wasn't good so far. I don't know yet how to debug them. But maybe it's time to learn something new :)

janiversen replied:
I use Visual Studio Code, when you open the workspace it tries “discover tests” when that runs successful you can run debug using the test entry.

vzahradnik replied:
Thanks for the tip. I use PyCharm, but I'm also familiar with VS Code. In recent weeks I use it every day for my Arduino projects in PlatformIO. I'll try your solution.

janiversen replied:
I used to use pycharm but switched to VC. Dif you add an extension to handle Arduino (I am no friend of the standard editor)?

vzahradnik replied:
Yes, the extension is called [PlatformIO IDE](https://platformio.org/). They support many frameworks and boards, Arduino is just one of them. However, most of my projects at the moment are for Arduino.

vzahradnik replied:
@janiversen I guess the only thing left in this PR is to write some unit tests for the Climate. I'll try to write some during this week. If I get stuck, I'll definitely ask for your help :)

janiversen replied:
If @MartinHjelmare is up to it, then let us get this PR merged! For sure I would like test cases but seeing this code merged is bigger priority.

@MartinHjelmare gave me a good idea on how to simplify the test cases quite a lot by using a pytest decorator. I am preparing a PR that updates all test cases.

vzahradnik replied:
OK, agree. While you'll work on test improvements, I will move to the other PR to get the work moving there too.

janiversen reviewed (changes requested):
I am happy with the code, it looks good. But I would really like to see the sensor test cases expanded.

vzahradnik commented:
> I am happy with the code, it looks good. But I would really like to see the sensor test cases expanded.

I will take a look at it.

vzahradnik committed:
Add custom data type support into Modbus climate
This commit adds support for custom data type and improves error handling.

Currently, if a user set data_count into a value other than 1, 2 or 4 bytes,
code threw an exception because it didn't support the parsing.

New behavior is to notify user that parsing was not successful,
and to inform him to specify custom data structure to use.

This behavior will be the same as used in Modbus sensor.

Fix linter warnings

vzahradnik committed:
Migrate Modbus climate to new schema

vzahradnik committed:
Fix exception due to not implemented method

MartinHjelmare commented on homeassistant/components/modbus/climate.py:
```diff
@@ -37,158 +45,131 @@
 
+    def set_hvac_mode(self, hvac_mode: str) -> None:
+        """Set new target hvac mode."""
+        self._hvac_mode = hvac_mode
```
This isn't used. Why do we implement this method?

vzahradnik replied:
Apparently, Home Assistant expects this method to be implemented. I noticed exceptions in a log, because Home Assistant was trying to call this method, which was not implemented.

While we support only AUTO HVAC mode, I think it is safe to keep the method.

MartinHjelmare replied:
What was calling this service?

vzahradnik replied:
![Snímka obrazovky z 2020-10-01 09-20-17](https://user-images.githubusercontent.com/4460766/94780616-090ba500-03c9-11eb-9eb7-dc00f2225713.png)

See the attached image. I think it is a button to change HVAC modes. In this case it calls the `set_hvac_mode()` method, but it will always pass the mode `AUTO`. This mode is the only one supported by the Modbus climate entity.

MartinHjelmare replied:
Ok, we can simplify here and just define the method without any function body.

vzahradnik replied:
OK, I'll update the code.

vzahradnik replied:
Code updated.

MartinHjelmare commented on homeassistant/components/modbus/climate.py:
```diff
@@ -37,158 +45,131 @@
+    if discovery_info is None:
+        return
+
+    climates = []
```
Let's rename this to `entities`.

vzahradnik replied:
OK, will do.

vzahradnik replied:
Done.

MartinHjelmare commented:
Please add a before and after example view of the configuration yaml section to the breaking changes paragrah.

vzahradnik commented:
> before and after example view

Done.

vzahradnik committed:
Rename climates array to entities

vzahradnik committed:
Rework set_hvac_mode()

Home Assistant requires this method; however, because our Climate
component doesn't support other modes than AUTO, we'll keep the method
body empty.

The only purpose is to avoid getting NotImplementedError due to this.

MartinHjelmare reviewed (approved):
Nice!

janiversen commented:
The failing test is not your problem. Simply push again (e.g. with an updated commit text)

vzahradnik committed:
Fix linter error

vzahradnik commented:
The checks were failing on the linter error - unnecessary `pass` statement. Everything is still working, and all the checks are green now. I think this PR is finally ready to be merged. Please let me know if there's anything left to do.

Thanks!

janiversen commented:
Interesting it worked fine in my environment (test), but super that you found it, and I see that @MartinHjelmare have approved the PR, so lets hope it gets merged soon.

vzahradnik commented:
> Interesting it worked fine in my environment (test)

It was a linter issue - `Unnecessary pass keyword`. Apparently, as long as you have some comment in a method, the `pass` keyword is not necessary. I didn't know that.

MartinHjelmare reviewed (approved):
Great!
